netsync/manager.go

```diff
@@ -895,12 +895,25 @@ func (sm *SyncManager) haveInventory(invVect *wire.InvVect) (bool, error) {
 		}
 
 		// Check if the transaction exists from the point of view of the
-		// end of the main chain.
-		entry, err := sm.chain.FetchUtxoEntry(&invVect.Hash)
+		// end of the main chain.  Note that this is only a best effort
+		// since it is expensive to check existence of every output and
+		// the only purpose of this check is to avoid downloading
+		// already known transactions.
+		prevOut := wire.OutPoint{Hash: invVect.Hash, Index: 0}
+		entry, err := sm.chain.FetchUtxoEntry(&prevOut)
 		if err != nil {
 			return false, err
 		}
-		return entry != nil && !entry.IsFullySpent(), nil
+		if entry != nil && !entry.IsSpent() {
+			return true, nil
+		}
+
+		prevOut.Index = 1
```

Comment:
Instead of having this block twice, what do you think of:

```go
prevOut := wire.OutPoint{Hash: invVect.Hash}
for i := 0; i < 2; i++ {
  prevOut.Index = i
  entry, err := sm.chain.FetchUtxoEntry(&prevOut)
  if err != nil {
    return false, err
  }
  if entry != nil && !entry.IsSpent() {
    return true, nil
  }
}

return false, nil
```

Reply:
That look good.  It should be a little more obvious what it's doing.  Changed.

I also updated the comment to provide some detail as to why per the conversation.